File: joss/comparisons/cat-4-join-safety/03-join-key/example-join-key.R
# Error Class 3: Join Key Errors
#
# R/dplyr join errors are caught at runtime. Missing keys produce errors,
# but some issues (like implicit cartesian joins from duplicate keys)
# silently inflate the result.

library(tidyverse)

patients <- read_csv("fixtures/patients.csv")
encounters <- read_csv("fixtures/encounters.csv")
labs <- read_csv("fixtures/lab_results.csv")

# ── ERROR 3a: Join key doesn't exist in one table ───────────────────────
# RUNTIME ERROR: Join columns must be present in data.
patients %>%
  left_join(labs, by = "encounter_id")
# Error: `encounter_id` not found in `x`

# ── ERROR 3b: Misspelled join key ───────────────────────────────────────
# RUNTIME ERROR: Column not found
patients %>%
  left_join(encounters, by = "patient_ID")

# ── ERROR 3c: Using columns from wrong table post-join ──────────────────
# RUNTIME ERROR: Column `prescription_id` doesn't exist.
patients %>%
  left_join(encounters, by = "patient_id") %>%
  mutate(note = prescription_id)
# Only caught when the pipeline actually runs

File: joss/comparisons/cat-4-join-safety/03-join-key/example-join-key.py
"""
Error Class 3: Join Key Errors

Python/pandas join errors are only caught at runtime. Misspelled keys
produce exceptions, but some mismatches produce silently wrong results
(e.g., joining on mismatched types creates an empty join).
"""
import pandas as pd

patients = pd.read_csv("fixtures/patients.csv")
encounters = pd.read_csv("fixtures/encounters.csv")
labs = pd.read_csv("fixtures/lab_results.csv")

# ── ERROR 3a: Join key doesn't exist in one table ───────────────────────
# RUNTIME: KeyError — only caught when this merge executes
merged = patients.merge(labs, on="encounter_id")
# KeyError: 'encounter_id' — patients has no encounter_id column

# ── ERROR 3b: Misspelled join key ───────────────────────────────────────
# RUNTIME: KeyError — case-sensitive mismatch
merged = patients.merge(encounters, on="patient_ID")
# KeyError: 'patient_ID'

# ── ERROR 3c: Using columns from wrong table post-join ──────────────────
# SILENT: After joining patients + encounters, you might reference
# a column from the medications table by mistake. Python doesn't
# track provenance — you only find out at runtime.
joined = patients.merge(encounters, on="patient_id")
try:
    result = joined["prescription_id"]  # KeyError at runtime
except KeyError:
    print("KeyError: 'prescription_id' — only discovered at runtime")
